mk add user model and validations

MelissaKaulfuss opened the pull request:
The branch adds:

- User model
- User model validations
- Tests (User factory and model specs)

This branch has no rubocop offences.

MelissaKaulfuss committed:
WIP: setup User model, controller

MelissaKaulfuss committed:
Uodate RSpec version to 3.7.1

MelissaKaulfuss committed:
Change Schema so that all User fields are not null

MelissaKaulfuss committed:
Fix typo

MelissaKaulfuss committed:
Fix typo

MelissaKaulfuss committed:
Add user spec

MelissaKaulfuss committed:
Changes RSpec version

MelissaKaulfuss committed:
WIP user spec

scara82 commented on config/routes.rb:
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
+  get 'users/new'
```
Maybe here you can use 
``resources :users, only: [:new]``
I just think it's nicer and you can easly add methods  ;)

MelissaKaulfuss replied:
i will :) thanks

scara82 commented on db/migrate/20180123050321_create_users.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :users do |t|
+      t.string :username
+      t.string :email
```
I think username and/or email should have
``null: false ``
to be sure all the users will enter one. (Don't they need a password as well? 🤔 )

MelissaKaulfuss commented:
Hmmm...so i did set them as not null in one of the commits on this branch (I hope!). Will have to have a look.

And yes, password digest is added in another branch to do with Authentication 👍

> On 27 Jan 2018, at 13:27, Marta <notifications@github.com> wrote:
> 
> @scara82 commented on this pull request.
> 
> In db/migrate/20180123050321_create_users.rb:
> 
> > @@ -0,0 +1,11 @@
> +class CreateUsers < ActiveRecord::Migration[5.1]
> +  def change
> +    create_table :users do |t|
> +      t.string :username
> +      t.string :email
> I think username and/or email should have
> null: false
> to be sure all the users will enter one. (Don't they need a password as well? 🤔 )
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

MelissaKaulfuss commented:
👍

> On 27 Jan 2018, at 13:19, Marta <notifications@github.com> wrote:
> 
> @scara82 commented on this pull request.
> 
> In config/routes.rb:
> 
> > @@ -1,3 +1,5 @@
>  Rails.application.routes.draw do
> +  get 'users/new'
> Maybe here you can use
> '''resources :users, only: [:new]'''
> I just think it's nicer and you can easly add methods ;)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

MelissaKaulfuss committed:
WIP re-edit user spec after Friday's change vanished

MelissaKaulfuss committed:
Change format of route for user#new action

MelissaKaulfuss committed:
WIP adding db/schema file

apalski commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,12 @@
+    end
+  end
+
+end
```
I used:
it { is_expected.to route(:get, "/users/new").to(controller: :users, action: :new) }
I will be interested to see if this is an acceptable format or whether I need the one you have used :)

apalski replied:
Elle, said the way I did the specs wasn't right for what I was doing, she pointed me back to the request specs we did in class

apalski replied:
Also, check Marta's PR for Elle's comments on where to put which type of specs for your controller

apalski commented on spec/factories/user_factory.rb:
```diff
@@ -0,0 +1,7 @@
+    email "felix@pip.com"
+    status 1
+  end
+end
```
I used:
factory :user do
    sequence(:name) { |n| "user name#{n}" }
    sequence(:email) { |n| "user email#{n}" }
  end
This was the way Elle had me set it out - she did suggest that I may want to have a check for the correct email format which I have yet to implement

apalski replied:
Everything looks good other than the two things Marta picked up - my comments are only for information - great work!

MelissaKaulfuss commented:
I honestly don't know why Status is still in there, that's the field I
removed from the schema when I dropped the table and redid the migration.
Weird!

Thanks :)

On 28 January 2018 at 16:10, Annette Drapalski <notifications@github.com>
wrote:

> *@apalski* commented on this pull request.
> ------------------------------
>
> In spec/factories/user_factory.rb
> <https://github.com/MelissaKaulfuss/stickit/pull/2#discussion_r164289215>:
>
> > @@ -0,0 +1,7 @@
> +FactoryBot.define do
> +  factory :user do
> +    username "felix_pip"
> +    email "felix@pip.com"
> +    status 1
> +  end
> +end
>
> Everything looks good other than the two things Marta picked up - my
> comments are only for information - great work!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MelissaKaulfuss/stickit/pull/2#discussion_r164289215>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACdXb-FHyN5vf7L9iWYvHmfx2kPMh2cYks5tPAFggaJpZM4RuNNN>
> .
>

MelissaKaulfuss committed:
Add User#create action

MelissaKaulfuss committed:
Add simple-form gem

MelissaKaulfuss committed:
Add responders gem

MelissaKaulfuss committed:
Add simple-form config

elle commented on config/initializers/simple_form.rb:
```diff
@@ -0,0 +1,169 @@
+  # config.input_class = nil
+
+  # Define the default class of the input wrapper of the boolean input.
+  config.boolean_label_class = 'checkbox'
```
Double quotes

elle commented on config/locales/simple_form.en.yml:
```diff
@@ -0,0 +1,31 @@
+    "no": 'No'
+    required:
+      text: 'required'
+      mark: '*'
```
Double quotes

elle commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,9 @@
+describe User do
+ context "validations" do
+
+    it { is_expected.to validate_presence_of(:username) }
```
Extra line

elle commented on app/models/user.rb:
```diff
@@ -0,0 +1,3 @@
+class User < ApplicationRecord
+  validates :username, :email, presence: true
```
A question: would username or email need to be unique?

MelissaKaulfuss replied:
Yes! you are absolutely right. I will add these constraints.

elle reviewed (commented):
Generally looks good.

- Missing feature specs to test that a new user can be created from the UI
- Adding the automatically generated new view with the default text is not useful
- The controller specs you have are testing that the routes are working. So:
  1. They belong in request specs, rather than controller specs
  2. They are not needed once you have feature specs in place

What I would do in the controller specs is check the sad path. For example, if you try to create a new user with invalid params, I would check that we get an error flash message, and can also see the errors.full_messages

elle reviewed (commented):
Also, add `*.DS_Store` to your `.gitignore`

elle commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,16 @@
+  end
+
+  def create
+   @user = User.new(user_params)
```
You are not creating a user here, just instantiating it.
Also, you are not rendering anything or returning anything.

elle commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+require "users_controller"
```
No need to manually call this line because the rails_helper loaded the whole rails environment

elle commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+require "users_controller"
+
+describe UsersController do
```
A suggestion to prefix this describe block with `RSpec.`

Ref: https://stackoverflow.com/questions/24747448/rspec-describe-vs-describe

elle commented on spec/factories/user_factory.rb:
```diff
@@ -0,0 +1,6 @@
+    username "felix_pip"
+    email "felix@pip.com"
+  end
+end
```
We prefer to use the single file for factories, because it makes it easier to track a factory down and see what is being set on which object.

elle commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,9 @@
+require "rails_helper"
+
+describe User do
```
Prefix `RSpec.`?

elle commented on Gemfile:
```diff
@@ -4,7 +4,9 @@ gem "jbuilder", "~> 2.5"
 gem "pg"
 gem "puma", "~> 3.7"
 gem "rails", "~> 5.1.4"
+gem "responders"
```
The responders gem is added but the install generator hasn't been run.
For future reference, I would add the gem, run the install command, and encapsulate those changes in one commit.

Unless the gem is being used, there is no reason to add it.

jonathanly commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Be wary of generated files, they tend to use single quotation marks.
It's caught me out a few times too!

elle commented on Gemfile:
```diff
@@ -13,7 +15,7 @@ group :development, :test do
   gem "hirb"
   gem "pry-byebug"
-  gem "rspec-rails"
+  gem "rspec-rails", "3.7.1"
```
I don't remember what issues we had with this version. What do you think about bumping it up to 3.7.2?

MelissaKaulfuss replied:
I've bumped this over in my other branch, see: fbdf31c6d0762e1a0e5e2b2d5bf90e8bad94543e

elle commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,2 @@
+<h1>Users#new</h1>
+<p>Find me in app/views/users/new.html.erb</p>
```
This view is the generated view. Currently, a user cannot be added via the UI because the form is not here.

MelissaKaulfuss replied:
I will make sure I don't commit these in the future. In the meantime, this is resolved over in my other branch (as it's being used now).

elle commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,18 @@
+    context "with valid attributes" do
+      it { is_expected.to route(:post, "/users")
+        .to(controller: :users, action: :create) }
+
```
Extra line
